lib/SILGen/SILGenPoly.cpp

```diff
@@ -3016,17 +3017,22 @@ SILGenFunction::createWithoutActuallyEscapingClosure(
     buildWithoutActuallyEscapingThunkBody(thunkSGF);
   }
 
-  CanSILFunctionType substFnType = thunkType->substGenericArgs(
-      F.getModule(), thunk->getForwardingSubstitutions());
+  CanSILFunctionType substFnTy = thunkType;
+  // Use the subsitution map in the context of the current function.
+  // thunk->getForwardingSubstitutionMap() / thunk might have been created in a
+  // different function's generic enviroment.
+  if (auto genericSig = thunkType->getGenericSignature()) {
+    substFnTy = thunkType->substGenericArgs(F.getModule(), interfaceSubs);
```

Comment:
Oh, nice catch. I wonder if thunks should always get a new generic environment instead of inheriting one from their parent. We might have the same issue with any other kind of thunk we emit lazily.

Reply:
On quick review, all the other places that call 'buildThunkType' perform this substitution.